File: firebase/functions/src/utils/apiCache.ts
/**
 * API Cache Utility for External Vendor APIs
 * 
 * Provides in-memory caching with TTL (Time To Live) for external API calls.
 * Reduces API rate limit issues and improves response times.
 * 
 * Usage:
 *   import { APICache, cachedRequest } from './utils/apiCache';
 *   
 *   const cache = new APICache(10); // 10 second default TTL
 *   
 *   // Use cached request wrapper
 *   const data = await cachedRequest(
 *     axios.get('https://api.example.com/data', { params }),
 *     cache,
 *     'https://api.example.com/data',
 *     params,
 *     apiKey
 *   );
 */

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import crypto from 'crypto';
import { URL } from 'url';

interface CacheEntry<T> {
  data: T;
  expiration: number;
}

export class APICache {
  private cache: Map<string, CacheEntry<any>>;
  private defaultTTL: number;
  private cleanupInterval: number;
  private lastCleanup: number;

  constructor(defaultTTL: number = 10, cleanupInterval: number = 60) {
    this.defaultTTL = defaultTTL;
    this.cleanupInterval = cleanupInterval;
    this.cache = new Map();
    this.lastCleanup = Date.now();
  }

  /**
   * Generate a cache key from URL, parameters, and API key.
   */
  generateKey(url: string, params?: Record<string, any>, apiKey?: string): string {
    try {
      // Normalize URL
      const urlObj = new URL(url);
      const baseUrl = `${urlObj.protocol}//${urlObj.host}${urlObj.pathname}`;
      
      // Merge URL query params with provided params
      const allParams: Record<string, string> = {};
      
      // Add URL query params
      urlObj.searchParams.forEach((value, key) => {
        allParams[key] = value;
      });
      
      // Add provided params
      if (params) {
        Object.keys(params).forEach(key => {
          allParams[key] = String(params[key]);
        });
      }
      
      // Sort parameters for consistent keys
      const sortedParams = Object.keys(allParams)
        .sort()
        .map(key => `${key}=${allParams[key]}`)
        .join('&');
      
      // Create key components
      const keyParts = [baseUrl, sortedParams];
      
      // Hash API key if provided
      if (apiKey) {
        const apiKeyHash = crypto.createHash('sha256').update(apiKey).digest('hex').substring(0, 16);
        keyParts.push(apiKeyHash);
      }
      
      // Generate final key
      const keyString = keyParts.join('|');
      return crypto.createHash('md5').update(keyString).digest('hex');
    } catch (error) {
      // Fallback if URL parsing fails
      const keyString = `${url}|${JSON.stringify(params || {})}|${apiKey || ''}`;
      return crypto.createHash('md5').update(keyString).digest('hex');
    }
  }

  /**
   * Get cached data if it exists and hasn't expired.
   */
  get<T>(key: string): T | null {
    // Periodic cleanup
    this.cleanupIfNeeded();
    
    const entry = this.cache.get(key);
    if (!entry) {
      return null;
    }
    
    // Check if expired
    if (Date.now() > entry.expiration) {
      this.cache.delete(key);
      return null;
    }
    
    console.debug(`Cache HIT for key: ${key.substring(0, 16)}...`);
    return entry.data as T;
  }

  /**
   * Store data in cache with TTL.
   */
  set<T>(key: string, value: T, ttl?: number): void {
    const ttlToUse = ttl || this.defaultTTL;
    const expiration = Date.now() + (ttlToUse * 1000);
    this.cache.set(key, { data: value, expiration });
    console.debug(`Cache SET for key: ${key.substring(0, 16)}... (TTL: ${ttlToUse}s)`);
  }

  /**
   * Delete a cache entry.
   */
  delete(key: string): void {
    if (this.cache.has(key)) {
      this.cache.delete(key);
      console.debug(`Cache DELETE for key: ${key.substring(0, 16)}...`);
    }
  }

  /**
   * Clear all cache entries.
   */
  clear(): void {
    this.cache.clear();
    console.info('Cache cleared');
  }

  /**
   * Remove expired entries if cleanup interval has passed.
   */
  private cleanupIfNeeded(): void {
    const currentTime = Date.now();
    if (currentTime - this.lastCleanup < this.cleanupInterval * 1000) {
      return;
    }
    
    this.lastCleanup = currentTime;
    const expiredKeys: string[] = [];
    
    this.cache.forEach((entry, key) => {
      if (currentTime > entry.expiration) {
        expiredKeys.push(key);
      }
    });
    
    expiredKeys.forEach(key => this.cache.delete(key));
    
    if (expiredKeys.length > 0) {
      console.debug(`Cleaned up ${expiredKeys.length} expired cache entries`);
    }
  }

  /**
   * Get cache statistics.
   */
  stats(): { totalEntries: number; activeEntries: number; expiredEntries: number; defaultTTL: number } {
    const currentTime = Date.now();
    let activeCount = 0;
    let expiredCount = 0;
    
    this.cache.forEach(entry => {
      if (currentTime > entry.expiration) {
        expiredCount++;
      } else {
        activeCount++;
      }
    });
    
    return {
      totalEntries: this.cache.size,
      activeEntries: activeCount,
      expiredEntries: expiredCount,
      defaultTTL: this.defaultTTL
    };
  }
}

/**
 * Wrapper function to make cached axios requests.
 * 
 * @param requestFn Function that returns an axios promise
 * @param cache APICache instance
 * @param url Request URL
 * @param params Request parameters
 * @param apiKey Optional API key
 * @param ttl Optional TTL override
 */
export async function cachedRequest<T>(
  requestFn: () => Promise<AxiosResponse<T>>,
  cache: APICache,
  url: string,
  params?: Record<string, any>,
  apiKey?: string,
  ttl?: number
): Promise<T> {
  // Generate cache key
  const cacheKey = cache.generateKey(url, params, apiKey);
  
  // Check cache first
  const cachedData = cache.get<T>(cacheKey);
  if (cachedData !== null) {
    return cachedData;
  }
  
  // Make API call
  const response = await requestFn();
  const data = response.data;
  
  // Cache the result
  cache.set(cacheKey, data, ttl);
  
  return data;
}

/**
 * Global cache instance (can be imported and used directly)
 */
let globalCache: APICache | null = null;

export function getCache(defaultTTL: number = 10): APICache {
  if (!globalCache) {
    globalCache = new APICache(defaultTTL);
  }
  return globalCache;
}

File: market_news_app/lib/utils/api_cache.dart
/// API Cache Utility for External Vendor APIs
/// 
/// Provides in-memory caching with TTL (Time To Live) for external API calls.
/// Reduces API rate limit issues and improves response times.
/// 
/// Usage:
///   final cache = APICache(defaultTTL: 10); // 10 second default TTL
///   
///   // Check cache first
///   final cachedData = cache.get<String>('cache_key');
///   if (cachedData != null) {
///     return cachedData;
///   }
///   
///   // Make API call
///   final data = await makeApiCall();
///   
///   // Cache the result
///   cache.set('cache_key', data, ttl: 10);
///   return data;

import 'dart:convert';
import 'dart:collection';

/// Cache entry storing data and expiration timestamp
class _CacheEntry<T> {
  final T data;
  final int expiration;

  _CacheEntry(this.data, this.expiration);

  bool get isExpired => DateTime.now().millisecondsSinceEpoch > expiration;
}

/// In-memory API cache with TTL support
class APICache {
  final int defaultTTL;
  final int cleanupInterval;
  
  final Map<String, _CacheEntry> _cache = {};
  int _lastCleanup = 0;

  APICache({this.defaultTTL = 10, this.cleanupInterval = 60});

  /// Generate a cache key from URL, parameters, and API key
  String generateKey(String url, Map<String, dynamic>? params, String? apiKey) {
    // Normalize URL (remove fragment)
    final uri = Uri.parse(url);
    final baseUrl = '${uri.scheme}://${uri.host}${uri.path}';
    
    // Merge URL query params with provided params
    final allParams = <String, String>{};
    
    // Add URL query params
    uri.queryParameters.forEach((key, value) {
      allParams[key] = value;
    });
    
    // Add provided params
    if (params != null) {
      params.forEach((key, value) {
        allParams[key] = value.toString();
      });
    }
    
    // Sort parameters for consistent keys
    final sortedKeys = allParams.keys.toList()..sort();
    final sortedParams = sortedKeys.map((key) => '$key=${allParams[key]}').join('&');
    
    // Create key components
    final keyParts = [baseUrl, sortedParams];
    
    // Hash API key if provided (simple hash for Dart)
    if (apiKey != null && apiKey.isNotEmpty) {
      final apiKeyHash = _hashString(apiKey).substring(0, 16);
      keyParts.add(apiKeyHash);
    }
    
    // Generate final key
    final keyString = keyParts.join('|');
    return _hashString(keyString);
  }

  /// Simple hash function for strings
  String _hashString(String input) {
    var bytes = utf8.encode(input);
    var hash = 0;
    for (var byte in bytes) {
      hash = ((hash << 5) - hash) + byte;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return hash.abs().toRadixString(16).padLeft(8, '0');
  }

  /// Get cached data if it exists and hasn't expired
  T? get<T>(String key) {
    // Periodic cleanup
    _cleanupIfNeeded();
    
    final entry = _cache[key];
    if (entry == null) {
      return null;
    }
    
    // Check if expired
    if (entry.isExpired) {
      _cache.remove(key);
      return null;
    }
    
    print('Cache HIT for key: ${key.substring(0, key.length > 16 ? 16 : key.length)}...');
    return entry.data as T;
  }

  /// Store data in cache with TTL
  void set<T>(String key, T value, {int? ttl}) {
    final ttlToUse = ttl ?? defaultTTL;
    final expiration = DateTime.now().millisecondsSinceEpoch + (ttlToUse * 1000);
    _cache[key] = _CacheEntry(value, expiration);
    print('Cache SET for key: ${key.substring(0, key.length > 16 ? 16 : key.length)}... (TTL: ${ttlToUse}s)');
  }

  /// Delete a cache entry
  void delete(String key) {
    if (_cache.containsKey(key)) {
      _cache.remove(key);
      print('Cache DELETE for key: ${key.substring(0, key.length > 16 ? 16 : key.length)}...');
    }
  }

  /// Clear all cache entries
  void clear() {
    _cache.clear();
    print('Cache cleared');
  }

  /// Remove expired entries if cleanup interval has passed
  void _cleanupIfNeeded() {
    final currentTime = DateTime.now().millisecondsSinceEpoch;
    if (currentTime - _lastCleanup < cleanupInterval * 1000) {
      return;
    }
    
    _lastCleanup = currentTime;
    final expiredKeys = <String>[];
    
    _cache.forEach((key, entry) {
      if (entry.isExpired) {
        expiredKeys.add(key);
      }
    });
    
    expiredKeys.forEach((key) => _cache.remove(key));
    
    if (expiredKeys.isNotEmpty) {
      print('Cleaned up ${expiredKeys.length} expired cache entries');
    }
  }

  /// Get cache statistics
  Map<String, dynamic> stats() {
    final currentTime = DateTime.now().millisecondsSinceEpoch;
    var activeCount = 0;
    var expiredCount = 0;
    
    _cache.forEach((key, entry) {
      if (entry.isExpired) {
        expiredCount++;
      } else {
        activeCount++;
      }
    });
    
    return {
      'totalEntries': _cache.length,
      'activeEntries': activeCount,
      'expiredEntries': expiredCount,
      'defaultTTL': defaultTTL,
    };
  }
}

/// Global cache instance (can be imported and used directly)
APICache? _globalCache;

APICache getCache({int defaultTTL = 10}) {
  _globalCache ??= APICache(defaultTTL: defaultTTL);
  return _globalCache!;
}
